src/ecdh.rs

```diff
@@ -132,11 +59,27 @@ impl SharedSecret {
         // The default `secp256k1_ecdh_hash_function_default` should always return 1.
         // and the scalar was verified to be valid(0 > scalar > group_order) via the type system
         debug_assert_eq!(res, 1);
-        ss.set_len(32); // The default hash function is SHA256, which is 32 bytes long.
-        ss
+        SharedSecret(buf)
+    }
+
+    /// Returns true if the underlying data buffer is empty.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
     }
 }
 
+impl AsRef<[u8]> for SharedSecret {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
```

Comment:
I think `Borrow` would be also useful.

Reply:
oh yes, this was discussed elsewhere already recently, my bad. Just to clarify, we should use `Borrow` and remove `AsRef`, right?

Reply:
We have `src/ecdsa/mod.rs:83:impl AsRef<[u8]> for SerializedSignature {`, should this be changed to implement `Borrow` instead?

Reply:
No, we should implement both `AsRef` and `Borrow` for shared secret.
For other types we have to evaluate them one-by-one since `Borrow<U> for T` requires that `T::Eq` and `T::Hash` behave exactly same as `U::Eq` and `U::Hash` provided the instance of `U` was obtained by calling `Borrow` on `T`.

Reply:
> No, we should implement both AsRef and Borrow for shared secret.

Cool, will add.

> since Borrow<U> for T requires that T::Eq and T::Hash behave exactly same as U::Eq and U::Hash provided the instance of U was obtained by calling Borrow on T.

I read that in the docs but I do not understand what it implies. I'll have to take direction from others on this until I do get it.

Reply:
Your code is correct. It would've been incorrect if `SharedSecret` overrode `Hash` or `Eq` implementation to behave differently than one on the slice it returns.